Indicators/ParabolicStopAndReverseExtended.cs

```diff
@@ -0,0 +1,304 @@
+                _outputSar = _sar;
+
+                // Adjust af and ep.
+                if (currentBar.High > _ep)
+                {
+                    _ep = currentBar.High;
+                    _afLong += _afIncrementLong;
+                    if (_afLong > _afMaxLong)
+                        _afLong = _afMaxLong;
+                }
+
+                // Calculate the new SAR
+                _sar = _sar + _afLong * (_ep - _sar);
+
+                // Make sure the new SAR is within yesterday's and today's range.
+                if (_sar > _previousBar.Low)
+                    _sar = _previousBar.Low;
+                if (_sar > currentBar.Low)
+                    _sar = currentBar.Low;
+            }
+        }
+
+        /// <summary>
+        /// Calculate indicator value when the position is short
+        /// </summary>
+        private void HandleShortPosition(IBaseDataBar currentBar)
+        {
+            // Switch to long if the high penetrates the SAR value.
+            if (currentBar.High >= _sar)
+            {
+                // Switch and overide the SAR with the ep
+                _isLong = true;
+                _sar = _ep;
+
+                // Make sure the overide SAR is within yesterday's and today's range.
+                if (_sar > _previousBar.Low)
+                    _sar = _previousBar.Low;
+                if (_sar > currentBar.Low)
+                    _sar = currentBar.Low;
+
+                // Output the overide SAR 
+                if(_offsetOnReverse != 0.0m)
+                    _sar -= _sar * _offsetOnReverse;
+                _outputSar = _sar;
+
+                // Adjust af and ep
+                _afLong = _afInitLong;
+                _ep = currentBar.High;
+
+                // Calculate the new SAR
+                _sar = _sar + _afLong * (_ep - _sar);
+
+                // Make sure the new SAR is within yesterday's and today's range.
+                if (_sar > _previousBar.Low)
+                    _sar = _previousBar.Low;
+                if (_sar > currentBar.Low)
+                    _sar = currentBar.Low;
+            }
+            //No switch
+            else
+            {
+                // Output the SAR (was calculated in the previous iteration)
+                _outputSar = -_sar;
+
+                // Adjust af and ep.
+                if (currentBar.Low < _ep)
+                {
+                    _ep = currentBar.Low;
+                    _afShort += _afIncrementShort;
+                    if (_afShort > _afMaxShort)
+                        _afShort = _afMaxShort;
+                }
+
+                // Calculate the new SAR
+                _sar = _sar + _afShort * (_ep - _sar);
+
+                // Make sure the new SAR is within yesterday's and today's range.
+                if (_sar < _previousBar.High)
+                    _sar = _previousBar.High;
+                if (_sar < currentBar.High)
+                    _sar = currentBar.High;
```

Comment:
These comparisons & assignments seem repeated a few times, seems they could be extracted into a private method to be reused? Will make the implementation easier to understand too I think